File: 345-Reverse-Vowels-of-a-String.java
class Solution {
      boolean isVowel(char c) {
        return c == 'a' || c == 'i' || c == 'e' || c == 'o' || c == 'u'
                || c == 'A' || c == 'I' || c == 'E' || c == 'O' || c == 'U';
    }

    void swap(char[] sArr, int start, int end) {
        char temp = sArr[start];
        sArr[start] = sArr[end];
        sArr[end] = temp;
    }

    public String reverseVowels(String s) {
        int start = 0, end = s.length() - 1;
        char[] sArr = s.toCharArray();

        while(start < end) {
            while(start < s.length() && !isVowel(sArr[start])) {
                start++;
            }
            while(end >= 0 && !isVowel(sArr[end])) {
                end--;
            }

            if(start < end) {
                swap(sArr, start++, end--);
            }
        }
        return new String(sArr);
    }
}

File: 345-Reverse-Vowels-of-a-String.go
func isVowel(c byte) bool {
\treturn c == 'a' || c == 'i' || c == 'e' || c == 'o' || c == 'u' ||
\t\tc == 'A' || c == 'I' || c == 'E' || c == 'O' || c == 'U'
}

func swap(chars []byte, start int, end int) {
    chars[start], chars[end] = chars[end], chars[start];
}

func reverseVowels(s string) string {
    start := 0;
    end := len(s) - 1;
    sChar := []byte(s);

    for start < end {
        for start < len(s) && !isVowel(sChar[start]) {
            start++;
        }
        for end >= 0 && !isVowel(sChar[end]) {
            end--;
        }
        if start < end {
            swap(sChar, start, end);
            start++;
            end--;
        }
    }
    return string(sChar);
}